### PR TITLE
Merge `main` into `berkeley` (routine)

### PR DESCRIPTION
**Pull request by eecavanna**
This is a routine merge of `main` into `berkeley`.

~GitHub is showing a "Can't automatically merge" alert, which I think means there will be some merge conflicts. I'll work through those after the PR gets created.~

See also: https://github.com/microbiomedata/nmdc-runtime/issues/577#issuecomment-2273969110

### CODE REVIEW

**Comment by eecavanna** [comment]
Here's the error message currently being shown in the GitHub Actions workflow [logs](https://github.com/microbiomedata/nmdc-runtime/actions/runs/10307783286/job/28533765354?pr=625#step:4:4099):
```py
E           fastapi.exceptions.HTTPException: 422: {'update_cmd': {'nmdc:sty-11-pzmd0x14': {'update': 'study_set', 'updates': [{'q': {'id': 'nmdc:sty-11-pzmd0x14'}, 'u': {'$set': {'type': 'nmdc:Study'}}}, {'q': {'id': 'nmdc:sty-11-pzmd0x14'}, 'u': {'$set': {'associated_dois': [{'doi_value': 'doi:10.25345/C5CG8S', 'doi_category': 'dataset_doi', 'doi_provider': 'massive'}]}}}]}}, 'validation_errors': [["'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property"]]}
```

It may be the case that I can't use the changesheet, itself, to set the `type` (but that I could, instead, do it in the database-seeding code in the test function, itself). I don't know.

**Comment by eecavanna** [comment]
Here's the error message currently being shown in the GitHub Actions workflow [logs](https://github.com/microbiomedata/nmdc-runtime/actions/runs/10307902922/job/28534154481?pr=625#step:4:4118):

```py
E           fastapi.exceptions.HTTPException: 422: {'update_cmd': {'nmdc:sty-11-pzmd0x14': {'update': 'study_set', 'updates': [{'q': {'id': 'nmdc:sty-11-pzmd0x14'}, 'u': {'$set': {'type': 'nmdc:Study'}}}, {'q': {'id': 'nmdc:sty-11-pzmd0x14'}, 'u': {'$set': {'associated_dois': [{'type': 'nmdc:Doi', 'doi_value': 'doi:10.25345/C5CG8S', 'doi_category': 'dataset_doi', 'doi_provider': 'massive'}]}}}]}}, 'validation_errors': [["'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property", "'type' is a required property"]]}
```

This is **after** I added these two lines to the changesheet.

```diff
  id	action	attribute	value
+ nmdc:sty-11-pzmd0x14	update	type	nmdc:Study
  nmdc:sty-11-pzmd0x14	insert	associated_dois	d1
+ d1	update	type	nmdc:Doi
  d1	update	doi_value	doi:10.25345/C5CG8S
  d1	update	doi_category	dataset_doi
  d1	update	doi_provider	massive
```

I suspect I cannot use a changesheet to a document that is already invalid (even if the changesheet's changes would make the document valid); in which case, it might be necessary to modify the code that inserts the documents into the database in the first place (i.e. the database "seeding" code).

**Comment by eecavanna** [comment]
I checked the file from which the initial Study gets inserted in the database (i.e. `tests/files/nmdc_sty-11-pzmd0x14.json`), and I see the Study already has its `type` field set correctly; so, I think adding the first new line to the changesheet was unnecessary.

I'll try to find the analogous file for the Doi instance.

**Comment by aclum** [comment]
@eecavanna I updated the test file with commit https://github.com/microbiomedata/nmdc-runtime/pull/625/commits/ec63e4af38b51b366b22b72c9d7b48a9cfe66147

**Comment by eecavanna** [comment]
Thanks, @aclum. I did something like that in https://github.com/microbiomedata/nmdc-runtime/pull/625/commits/51bd7435ff592d50c955e41d754f1a58e8c086ca and then undid it after concluding it had no effect on the result.

**Comment by eecavanna** [comment]
I'll be offline from 5:45-9:30pm PT. You can continue making changes on this branch (even trial and error ones) throughout the day today. I can always recreate the final changes in a new PR, which will have a less "noisy" commit history.

**Comment by aclum** [comment]
@eecavanna  this is now failing at the expected place, test_gold_translator.py

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (aclum) [inline docs/nb/bulk_validation_referential_integrity_check.ipynb:351]: This doesn't need to be addressed in this PR but this else logic can be removed for berkeley, type is required and collection names now in some cases refer to abstract classes. @PeopleMakeCulture

**Comment by eecavanna** [comment]
Awesome—thanks, @aclum! I am going to try to condense all those commits into 1-2, while preserving you as the author.

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (aclum) [inline docs/nb/bulk_validation_referential_integrity_check.ipynb:652]: Activity here should be changed to DataGeneration

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (aclum) [inline docs/nb/bulk_validation_referential_integrity_check.ipynb:688]: OmicsProcessing should be DataGeneration.

File: nmdc_runtime/site/export/ncbi_xml_utils.py
Comment (aclum) [inline nmdc_runtime/site/export/ncbi_xml_utils.py:120]: change designated_class to `type`

**Review by aclum — APPROVED** [review]
I made some individual comments but those are all for scripts that haven't been updated for berkeley yet so don't need to be resolved before this PR is merged, we should go ahead and merge this in so people can update with the latest code from main.
cc @sujaypatil96 @PeopleMakeCulture

**Comment by eecavanna** [comment]
Thanks, @aclum! I will merge this in once the latest GitHub Actions workflow finishes running (I'll confirm it fails in the "expected" place still, now that I've "condensed" all the recent commits into a single one).

I'll create individual issues for the things you pointed out in the comments above.

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (eecavanna) [reply docs/nb/bulk_validation_referential_integrity_check.ipynb:652]: I created issue https://github.com/microbiomedata/nmdc-runtime/issues/628 about this.

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (eecavanna) [reply docs/nb/bulk_validation_referential_integrity_check.ipynb:688]: I created issue https://github.com/microbiomedata/nmdc-runtime/issues/629 about this.

File: nmdc_runtime/site/export/ncbi_xml_utils.py
Comment (eecavanna) [reply nmdc_runtime/site/export/ncbi_xml_utils.py:120]: I created issue https://github.com/microbiomedata/nmdc-runtime/issues/630 about this.

File: docs/nb/bulk_validation_referential_integrity_check.ipynb
Comment (eecavanna) [reply docs/nb/bulk_validation_referential_integrity_check.ipynb:351]: I created a standalone ticket about this topic. Ticket: https://github.com/microbiomedata/nmdc-runtime/issues/632

**Pull request merged**
